### DISCUSSION TITLE
How do I get inode in `getattr` if chached in `readdir`?

### DISCUSSION THREAD

**Discussion by GiorgosXou** [libfuse/libfuse#1035] [General]
Say I have:
```c
    ...
    cfg->readdir_ino=1;
    cfg->remember=-1;
    cfg->kernel_cache = 1;
```
```c
static int my_readdir(const char *path, void *buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info *fi, enum fuse_readdir_flags flags) {
	if (strcmp(path, "/") != 0)
		return -ENOENT;

	printf("readdir: %s\n", path );
	filler(buf, ".", NULL, 0, 0);
	filler(buf, "..", NULL, 0, 0);

	struct stat st;
	memset(&st, 0, sizeof(st));
	st.st_ino = 9465405; // NOTE: INODE CACHED ???????
	st.st_mode = S_IFREG | 0444; 
	filler(buf, "hello.txt", &st, 0, 0); // // maybe use "..., FUSE_FILL_DIR_PLUS);"
	return 0;
}
```
How am I supposed to get the cached-inode read at `my_getattr(...)`? is this even possible?

**Comment by trapexit** [comment]
Not sure what you are asking. You are using the high level API and as such using a path based API. It is on you to orchestrate the path meaning.

**Reply by GiorgosXou** [reply]
> Not sure what you are asking.

I'm asking if it's possible to fill the **inode** during `readdir` via `filler`, such that I can access **it** later during `gettatr` *(without the need of any external mapping-mechanism of paths-to-inodes)*, the same way I'm able to access `path`s, once and for all.

> You are using the high level API

I tried using `fuse_lowlevel.h` too but I had an issue of the same nature.

**Reply by trapexit** [reply]
No, the inode value is no different from the size or mode. You are responsible for managing it.

**Reply by GiorgosXou** [reply]
Then what's the purpose of `readdir_ino`, `remember` and `kernel_cache`? why do they cache it if there's no way of accessing?

**Reply by trapexit** [reply]
I'm not following. Caching is for the client. Not the server. You are developing the server. You are responsible for managing the data. The inoded is just another point of data like all others in getattr you are expected to provide. I don't understand what you mean by "accessing." You created the data in the first place.

**Reply by GiorgosXou** [reply]
See the intro of  [this](https://stackoverflow.com/questions/78986903) *(if it makes any sense at all)*. Basically I'm just trying to "map" a few inodes from an external FS to a VFS. That's why i was wondering if there's any way of keeping track of inodes without any external mechanism.

**Reply by GiorgosXou** [reply]
Thanks though for clarifying caching, now it makes sense.

**Reply by trapexit** [reply]
I'm still not sure I understand what you are expecting to exist. You are responsible for building out the filesystem tree. You are responsible for filling in relevant entry and attr data. The inode value in attr is just a random value you decide to put there. It means absolutely nothing to the kernel. If you want for that value to have meaning you have to associate the path (or node,gen if using low level api) to whatever file/data you care about.

**Reply by GiorgosXou** [reply]
> Caching is for the client. 

Considering that the "client" is caching, is there any way of accessing this cached data from the "client"? 
*<sup>(Sorry for bothering you, last question.)</sup>*

**Reply by trapexit** [reply]
The client isn't caching. The caching is *for* the client. Server == FUSE filesystem. Client == application using the filesystem.

**Reply by bsbernd** [reply]
@trapexit Actually with kernel we now use " Server == FUSE filesystem server (daemon). Client == fuse-kernel and Application == part that accesses the vfs mount". Personally I use "FUSE-server", as for my typical use case the server is on the other side of the network.

**Reply by bsbernd** [reply]
@GiorgosXou all the caching meaning is for the kernel (as stated above "client"), you now need to update your question with these terms. And no, you cannot access kernel cache info from fuse-server/daemon - if you want to do that, it is your responsibility.

**Reply by trapexit** [reply]
The kernel being the technical client isn't really meaningful in these discussions. The kernel is effectively always the client. From the userland dev and end user perspective it's just a middle man. What is important in this discussion is that as the server... it is the dev's responsibility to manage data.

**Reply by GiorgosXou** [reply]
> you now need to update your question with these terms.

I voted to close my question. I ended up "solving" my issue by populating filenames via `filler` with their corresponding inode from the external filesystem such that I could access those files later via their inodes + `libext2fs`. *(like: "filename_\<inode\>" [max 255 chars of course])*

>  if you want to do that, it is your responsibility.

As of now I'm pretty satisfied with my workaround, but thanks for clarifying.

**Comment by bsbernd** [comment]
I don't think you need libext2fs, but just name_to_handle_at/open_by_handle_at.

**Reply by GiorgosXou** [reply]
If that's true it's solves alot for me. So... If all the attributes set to a virtual-file *(via `getattr` and etc.)* are exactly eqivalent to a file i'm trying to access from an external file-system, then: I am able to access somehow this file using it's virtual-path  + `name_to_handle_at/open_by_handle_at`?

**Comment by bsbernd** [comment]
`name_to_handle_at` gives you a handle to the file inode. `open_by_handle_at` can open that handle, without the need for a path.

**Reply by GiorgosXou** [reply]
Ok, but `name_to_handle_at` needs a `pathname`.  How am I supposed to use the virtual-pathname to get the handle for a file that belongs to an external file-system?

**Reply by bsbernd** [reply]
I would call name_to_handle_at in the lookup call, store it into my file systems inode and then access that (from the nodeid) in the calls you need it (like read/write).

**Reply by GiorgosXou** [reply]
@bsbernd **You were right!!**  Somehow I missed the fact that file-handles don't get stale after reboot! you saved me from a huge rabithole, **thank you so much!**

**Comment by bsbernd** [comment]
@trapexit maybe not meaningful, but I get really confused myself if everyone use different terms. I guess we should add to the docs, what is the definition of client and server...
